subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java

```diff
@@ -44,14 +43,13 @@ public ImplementationChanges(
 
     @Override
     public boolean accept(ChangeVisitor visitor) {
+        // In this method, the current implementation and additional implementations can't be unknown, since that causes already a validation warning.
+        // Previous implementations can still be unknown, since we store the inputs in the task history even if validation fails.
+        // When we fail the build for unknown implementations, then the previous implementations also can't be unknown.
```

Comment:
This would be so much better to capture by using the type system, say, by introducing something like `PotentiallyUnknownImplementationSnapshot`, or something.

Reply:
I did a change in that direction.